packages/auth/src/client/service.ts

```diff
@@ -166,14 +297,77 @@ function verifyClientDisplay(
   )
 }
 
+function isJwkViable(keys: RegistryKey): boolean {
+  return !!(
+    (!keys.exp || new Date() < new Date(keys.exp * 1000)) &&
+    (!keys.nbf || new Date() >= new Date(keys.nbf * 1000))
+  )
+}
+
 function verifyJwk(jwk: JWKWithRequired, keys: RegistryKey): boolean {
   // TODO: update this to reflect eventual shape of response from registry
-  return !!(
-    !keys.revoked &&
-    keys.exp &&
-    new Date() < new Date(keys.exp * 1000) &&
-    keys.nbf &&
-    new Date() >= new Date(keys.nbf * 1000) &&
-    keys.x === jwk.x
+  return !!(!keys.revoked && isJwkViable(keys) && keys.x === jwk.x)
+}
+
+async function tokenHttpsigMiddleware(
+  deps: ServiceDependencies,
+  ctx: AppContext,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  next: () => Promise<any>
+): Promise<void> {
+  const sig = ctx.headers['signature']
+  const sigInput = ctx.headers['signature-input']
+
+  if (
+    !sig ||
+    !sigInput ||
+    typeof sig !== 'string' ||
+    typeof sigInput !== 'string'
+  ) {
+    ctx.status = 400
+    ctx.body = {
+      error: 'invalid_request',
+      message: 'invalid signature headers'
+    }
+    next()
+    return
+  }
+
+  const { body } = ctx.request
+  let keyName = '',
+    value = ''
+  const { path, method } = ctx
+  // TODO: replace with HttpMethod types instead of string literals
+  if (path.includes('/introspect') && method === 'POST') {
+    keyName = 'value'
+    value = body['access_token']
+  } else if (path.includes('/token') && method === 'DELETE') {
+    keyName = 'managementId'
+    value = ctx.params['managementId']
+  } else {
+    // Is not a route that requires httpsig validation, somehow
+    next()
```

Comment:
Can we ditch `useHttpSigMiddleware` and use this middleware for all routes if it's making determinations of what requires httpsig validation?